<p align="center">
    <img width="50%" src="https://kaopubear-1254299507.cos.ap-shanghai.myqcloud.com/picgo/20210401150558.png">
</p>

**Triti-Map** is a Snakemake-based pipeline for gene mapping in Triticeae, which contains a suite of user-friendly computational packages and web-interface integrating multi-omics data from Triticeae species including genomic, epigenomic, evolutionary and homologous information.

**Triti-Map** could efficiently explore trait-related genes or functional elements not present in the reference genome and reduce the time and labor required for gene mapping in large genome species.

**More thorough information and explanations are provided in the [Triti-Map Manual](https://github.com/fei0810/Triti-Map/wiki/manual).**

### Triti-Map workflow overview

<p align="center">
    <img width="80%" src="https://kaopubear-1254299507.cos.ap-shanghai.myqcloud.com/picgo/20210330151508.png">
</p>

**Optimization steps to adress specific challengens of Triticeae gene-mapping**

<p align="center">
    <img width="80%" src="https://kaopubear-1254299507.cos.ap-shanghai.myqcloud.com/picgo/20210330151509.png">
</p>

## Citing
